vm/src/deserializer.h

```diff
@@ -0,0 +1,83 @@
+#ifndef BORGES_DESERIALIZER_H
+#define BORGES_DESERIALIZER_H
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace Borges {
+
+class Deserializer {
+public:
+    Deserializer(std::size_t n, const double* p);
+    void deserialize(double* p);
+    template <typename T>
+    void deserialize(std::shared_ptr<T>* p);
+
+private:
+    std::size_t size_;
+    const double* data_;
+    std::size_t index_;
+    std::unordered_map<std::size_t, std::shared_ptr<void>> index_to_ptr_map_;
+};
+
+void deserialize(Deserializer& d, bool* p);
+void deserialize(Deserializer& d, double* p);
+void deserialize(Deserializer& d, std::size_t* p);
+
+template <typename T>
+T
+deserialize(Deserializer& d)
+{
+    alignas(T) char b[sizeof(T)];
+    T* p = reinterpret_cast<T*>(b);
+    deserialize(d, p);
+    return *p;
+}
```

Comment:
aren't you returning a value on the stack here ?

Reply:
This is a bit tricky.

As I explained earlier, we create a buffer b to uninitialized storage here, and then take a pointer p to that storage. The call to deserialize on line 35 deserializes and constructs a value in the uninitialized storage pointed to by p. In other words, after this call, p now points to a value of type T, which is stored in the buffer b. When we return *p, what we actually return is a *copy* of the value pointed to by p.

However, you did make me realise a subtle bug in this code. When we return from this function, the buffer b goes out of scope, but we never call the destructor for the value that it stores. That's a potential memory leak if the value in question owns any pointers that it needs to release in it's destructor.

To avoid this, we should first copy the value stored in the buffer b to a temporary value v, explicitly call the destructor for the value in the buffer, and *then* return a copy of v (unlike the value stored in but, the destructor for v will be called automatically when we go out of scope).

As an extra optimization, we should use move instead of copy semantics here to avoid potentially expensive copying. I will update the pull request accordingly.